File: eHour-persistence/src/main/java/net/rrm/ehour/persistence/activity/dao/ActivityDaoHibernateImpl.java
/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

package net.rrm.ehour.persistence.activity.dao;

import java.util.List;

import net.rrm.ehour.domain.Activity;
import net.rrm.ehour.domain.Project;
import net.rrm.ehour.domain.User;
import net.rrm.ehour.persistence.dao.AbstractGenericDaoHibernateImpl;

import org.springframework.stereotype.Repository;

@Repository("activityDao")
public class ActivityDaoHibernateImpl extends AbstractGenericDaoHibernateImpl<Activity, Integer> implements ActivityDao {

	public ActivityDaoHibernateImpl() {
		super(Activity.class);
	}

	@SuppressWarnings("unchecked")
	@Override
	public List<Activity> findAllActivitiesOfUser(User assignedUser) {
		return getHibernateTemplate().findByNamedQueryAndNamedParam("Activity.findByUser", "assignedUser", assignedUser);
	}

	@SuppressWarnings("unchecked")
	@Override
	public List<Activity> findAllActivitiesOfProject(Project project) {
		return getHibernateTemplate().findByNamedQueryAndNamedParam("Activity.findByProject", "project", project);
	}

}

File: eHour-persistence/src/main/scala/net/rrm/ehour/persistence/activity/dao/ActivityDaoHibernateImpl.scala
package net.rrm.ehour.persistence.activity.dao

import java.util

import com.google.common.collect.Lists
import net.rrm.ehour.data.DateRange
import net.rrm.ehour.domain.{Activity, Customer, Project, User}
import net.rrm.ehour.persistence.dao.AbstractGenericDaoHibernateImpl
import org.springframework.stereotype.Repository

@Repository("activityDao")
class ActivityDaoHibernateImpl extends AbstractGenericDaoHibernateImpl[Integer, Activity](classOf[Activity]) with ActivityDao {
  final val CacheRegion = Some("query.Activity")

  override def findAllActivitiesOfUser(assignedUser: User): util.List[Activity] = findByNamedQuery("Activity.findByUser", "assignedUser", assignedUser)

  override def findAllActivitiesOfProject(project: Project): util.List[Activity] = findByNamedQuery("Activity.findByProject", "project", project)

  override def findActivitiesOfProject(project: Project, dateRange: DateRange) = {
    val keys  = List("project", "dateStart", "dateEnd")
    val params  = List(project, dateRange.getDateStart, dateRange.getDateEnd)

    findByNamedQuery("Activity.findByProjectAndDateRange", keys, params, CacheRegion)

  }

  override def findActivitiesForUser(userId: Integer, dateRange: DateRange): util.List[Activity] = {
    val keys  = List("dateStart", "dateEnd", "userId")
    val params  = List(dateRange.getDateStart, dateRange.getDateEnd, userId)

    findByNamedQuery("Activity.findActivitiessForUserInRange", keys, params, CacheRegion)
  }

  override def findActivitiesForCustomers(customers: util.List[Customer]): util.List[Activity] = {
    if (customers == null || customers.isEmpty) {
      return Lists.newArrayList()
    }

    findByNamedQuery("Activity.findActivitiesForCustomers", "customers", customers, CacheRegion)
  }

  override def findActivitiesForCustomers(customers: util.List[Customer], dateRange: DateRange): util.List[Activity] = {
    val keys  = List("customers", "dateStart", "dateEnd")
    val params  = List(customers, dateRange.getDateStart, dateRange.getDateEnd)

    findByNamedQuery("Activity.findActivitiessForUserInRange", keys, params, CacheRegion)
  }

  /**
   * Searches and if found returns an {@link Activity} having the code same as
   * passed in the parameter.
   */
  override def findByCode(code: String): Activity = {
    val xs = findByNamedQuery("Activity.findByCode","code", code, CacheRegion)

    if (xs.isEmpty) null else xs.get(0)
  }
}
